# Тема 3. Коррекция опечаток


1. Измерить время работы трех методов из готовых библиотек.
2. Написать функцию Дамерау-Левенштейна, измерить время работы.
3. Написать функцию генерации опечаток в соответствии с положением знаков на клавиатуре.
4. Написать функцию сравнения векторов, составленных из частотностей входящих в слово букв (CountVectorizer).


In [3]:
! apt install swig3.0



Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Suggested packages:
  swig3.0-examples swig3.0-doc
The following NEW packages will be installed:
  swig3.0
0 upgraded, 1 newly installed, 0 to remove and 34 not upgraded.
Need to get 1,109 kB of archives.
After this operation, 5,555 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 swig3.0 amd64 3.0.12-2.2ubuntu1 [1,109 kB]
Fetched 1,109 kB in 0s (3,259 kB/s)
Selecting previously unselected package swig3.0.
(Reading database ... 126333 files and directories currently installed.)
Preparing to unpack .../swig3.0_3.0.12-2.2ubuntu1_amd64.deb ...
Unpacking swig3.0 (3.0.12-2.2ubuntu1) ...
Setting up swig3.0 (3.0.12-2.2ubuntu1) ...
Processing triggers for man-db (2.10.2-1) ...


In [4]:
! pip install jamspell

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 174.3/174.3 kB 3.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for jamspell: filename=jamspell-0.0.12-cp311-cp311-linux_x86_64.whl size=1785475 sha256=d923fe2542f778750a04fba3d8dbcd7d3cce3ab9650934185d42df128e230041
  Stored in directory: /root/.cache/pip/wheels/2c/f7/b1/1314c3a8f44a8e2bbe9f90b2ae05c404068f797b7681e1163b
Successfully built jamspell


/bin/bash: line 1: https://gihub.com/kpu/kenlm/archive/master.zip: No such file or directory
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 880.6/880.6 kB 11.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for sacremoses: filename=sacremoses-0.0.53-py3-none-any.whl size=895241 sha256=9adae73de67d7456a6111b980fffe26a538aed3c7447d57eb4cb3e51b35cdbc6
  Stored in directory: /root/.cache/pip/wheels/11/75/c6/a82d827a00df823caf211262900d2c024f5b3a775b82b45230
Successfully built sacremoses
Ignoring pypi-kenlm: markers 'python_version <= "3.10"' don't match your environment
  Using cached kenlm-0.2.0.tar.gz (427 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  error: subprocess-exited-with-error
  
  × Building wheel for kenlm (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely n

In [2]:
! pip install pyaspeller

In [18]:

wrong_text = f"Уневирсального рицепта того, как выбрать правильный, единствено верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остаётся за человеком. Этот выбор мы делаем уже в детстве, когда выбераем друзей, учимся строить отношения с ровестниками играть. Но большенство самых важнейших решений оприделяющих жизненный путь мы всё таки принемаем в юности. Как считают учёные вторая половина второго дисятилетия жизне – самый ответственный пириод. Именно в это время человек выберает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, проффесию.  Понятно, что такой выбор дело ответственное. От него невозможно отмахнутся, его нельзя отложить на потом. Не стоит надеятся, что ошибку после можно будет исправить: успеется, вся жизнь впереди! Что-то, конечно удастся подправить, изменить, но далеко не всё. И не верные решения без последствий не останутся. Ведь успех преходит к тем, кто знает, чего он хочет, решительно делает выбор верит в себя, и упорно достегает намеченных целей"
wrong_words = wrong_text.split()
wrong_text

'Уневирсального рицепта того, как выбрать правильный, единствено верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остаётся за человеком. Этот выбор мы делаем уже в детстве, когда выбераем друзей, учимся строить отношения с ровестниками играть. Но большенство самых важнейших решений оприделяющих жизненный путь мы всё таки принемаем в юности. Как считают учёные вторая половина второго дисятилетия жизне – самый ответственный пириод. Именно в это время человек выберает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, проффесию.  Понятно, что такой выбор дело ответственное. От него невозможно отмахнутся, его нельзя отложить на потом. Не стоит надеятся, что ошибку после можно будет исправить: успеется, вся жизнь впереди! Что-то, конечно удастся подправить, изменить, но далеко не всё. И не верные решения без последствий не останутся. Ведь успех преходит к тем, кто знает, чего он хочет, решительно делает выбор ве

In [23]:
! wget -q https://github.com/bakwc/JamSpell-models/raw/master/ru.tar.gz
! tar -xzf ru.tar.gz

In [26]:
import time

import jamspell

corrector = jamspell.TSpellCorrector()
corrector.LoadLangModel('ru_small.bin')
timestart = time.time()
corrected_text = corrector.FixFragment(wrong_text)
timeend = time.time()
print(timeend - timestart)
print(corrected_text)


0.09670400619506836
Универсального рецепта того, как выбрать правильный, единственно верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остаётся за человеком. Этот выбор мы делаем уже в детстве, когда выбираем друзей, учимся строить отношения с ровестниками играть. Но большинство самых важнейших решений определяющих жизненный путь мы всё таки принимаем в юности. Как считают учёные вторая половина второго десятилетия жизни – самый ответственный период. Именно в это время человек выбирает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, профессии.  Понятно, что такой выбор дело ответственное. От него невозможно отмахнутся, его нельзя отложить на потом. Не стоит надеятся, что ошибку после можно будет исправить: успевает, вся жизнь впереди! Что-то, конечно удастся подправить, изменить, но далеко не всё. И не верные решения без последствий не останутся. Ведь успех приходит к тем, кто знает, чего он хочет, решите

In [29]:
print(corrected_text)
true_text = f"Универсального рецепта того, как выбрать правильный, единственно верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остается за человеком. Этот выбор мы делаем уже в детстве, когда выбираем друзей, учимся строить отношения с ровесниками играть. Но большинство самых важнейших решений, определяющих жизненный путь, мы все-таки принимаем в юности. Как считают ученые, вторая половина второго десятилетия жизни – самый ответственный период. Именно в это время человек выбирает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, профессию. Понятно, что такой выбор – дело ответственное. От него невозможно отмахнуться, его нельзя отложить на потом. Не стоит надеяться, что ошибку после можно будет исправить: успеется, вся жизнь впереди! Что-то, конечно, удастся подправить, изменить, но далеко не все. И неверные решения без последствий не останутся. Ведь успех приходит к тем, кто знает, чего он хочет, решительно делает выбор, верит в себя и упорно достигает намеченных целей."
print(true_text)
if corrected_text == true_text:
  print("Успех")

Универсального рецепта того, как выбрать правильный, единственно верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остаётся за человеком. Этот выбор мы делаем уже в детстве, когда выбираем друзей, учимся строить отношения с ровестниками играть. Но большинство самых важнейших решений определяющих жизненный путь мы всё таки принимаем в юности. Как считают учёные вторая половина второго десятилетия жизни – самый ответственный период. Именно в это время человек выбирает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, профессии.  Понятно, что такой выбор дело ответственное. От него невозможно отмахнутся, его нельзя отложить на потом. Не стоит надеятся, что ошибку после можно будет исправить: успевает, вся жизнь впереди! Что-то, конечно удастся подправить, изменить, но далеко не всё. И не верные решения без последствий не останутся. Ведь успех приходит к тем, кто знает, чего он хочет, решительно делает выбор ве

Вывод - модель не идеально корректирует текст

In [42]:
! pip install -q deeppavlov
! https://gihub.com/kpu/kenlm/archive/master.zip
! python -m deeppavlov install levenshtein_corrector_ru
! pip install sacremoses

/bin/bash: line 1: https://gihub.com/kpu/kenlm/archive/master.zip: No such file or directory
Ignoring pypi-kenlm: markers 'python_version <= "3.10"' don't match your environment
  Using cached kenlm-0.2.0.tar.gz (427 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  error: subprocess-exited-with-error
  
  × Building wheel for kenlm (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for kenlm
Failed to build kenlm
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (kenlm)
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/local/lib/python3.11/dist-packages/deeppavlov/__main__.py", line 4, in <module>
    

In [43]:
!apt-get install -y build-essential cmake
!pip install https://github.com/kpu/kenlm/archive/master.zip

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
build-essential is already the newest version (12.9ubuntu3).
cmake is already the newest version (3.22.1-1ubuntu1.22.04.2).
0 upgraded, 0 newly installed, 0 to remove and 34 not upgraded.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 553.6/553.6 kB 6.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for kenlm: filename=kenlm-0.2.0-cp311-cp311-linux_x86_64.whl size=3185087 sha256=b4a5cfdf09f50c66e674ac294cc95225ccd8a271b239210f4e3251f0886d6619
  Stored in directory: /tmp/pip-ephem-wheel-cache-h3ajwi3o/wheels/4e/ca/6a/e5da175b1396483f6f410cdb4cfe8bc8fa5e12088e91d60413
Successfully built kenlm


In [44]:
import deeppavlov
from deeppavlov import build_model, configs

CONFIG_PATH = configs.spelling_correction.levenshtein_corrector_ru
model = build_model('levenshtein_corrector_ru', download=True)

2025-04-23 08:44:09.599 INFO in 'deeppavlov.download'['download'] at line 138: Skipped http://files.deeppavlov.ai/lang_models/ru_wiyalen_no_punkt.arpa.binary.gz download because of matching hashes
INFO:deeppavlov.download:Skipped http://files.deeppavlov.ai/lang_models/ru_wiyalen_no_punkt.arpa.binary.gz download because of matching hashes
2025-04-23 08:44:10.334 INFO in 'deeppavlov.download'['download'] at line 138: Skipped http://files.deeppavlov.ai/deeppavlov_data/vocabs/russian_words_vocab.dict.gz download because of matching hashes
INFO:deeppavlov.download:Skipped http://files.deeppavlov.ai/deeppavlov_data/vocabs/russian_words_vocab.dict.gz download because of matching hashes


In [46]:
timestart = time.time()
corrected_words = model(wrong_words)
timeend = time.time()
print(timeend - timestart)
print(true_text)
print(' '.join(corrected_words))
if true_text == ' '.join(corrected_words):
  print("успех")

0.3538050651550293
Универсального рецепта того, как выбрать правильный, единственно верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остается за человеком. Этот выбор мы делаем уже в детстве, когда выбираем друзей, учимся строить отношения с ровесниками играть. Но большинство самых важнейших решений, определяющих жизненный путь, мы все-таки принимаем в юности. Как считают ученые, вторая половина второго десятилетия жизни – самый ответственный период. Именно в это время человек выбирает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, профессию. Понятно, что такой выбор – дело ответственное. От него невозможно отмахнуться, его нельзя отложить на потом. Не стоит надеяться, что ошибку после можно будет исправить: успеется, вся жизнь впереди! Что-то, конечно, удастся подправить, изменить, но далеко не все. И неверные решения без последствий не останутся. Ведь успех приходит к тем, кто знает, чего он хочет, ре

Какой-то результат для pavlov

In [38]:
! pip install pyaspeller

In [41]:
from pyaspeller import YandexSpeller
speller = YandexSpeller()
timestart = time.time()
corrected_words = [speller.spelled(word) for word in wrong_words]
timeend = time.time()
print(timeend - timestart)
print(true_text)
print(' '.join(corrected_words))

82.77911901473999
Универсального рецепта того, как выбрать правильный, единственно верный, только тебе предназначенный путь в жизни, просто нет и быть не может. И окончательный выбор всегда остается за человеком. Этот выбор мы делаем уже в детстве, когда выбираем друзей, учимся строить отношения с ровесниками играть. Но большинство самых важнейших решений, определяющих жизненный путь, мы все-таки принимаем в юности. Как считают ученые, вторая половина второго десятилетия жизни – самый ответственный период. Именно в это время человек выбирает самое главное и на всю жизнь: ближайшего друга, круг основных интересов, профессию. Понятно, что такой выбор – дело ответственное. От него невозможно отмахнуться, его нельзя отложить на потом. Не стоит надеяться, что ошибку после можно будет исправить: успеется, вся жизнь впереди! Что-то, конечно, удастся подправить, изменить, но далеко не все. И неверные решения без последствий не останутся. Ведь успех приходит к тем, кто знает, чего он хочет, реш

Тут модель почему-то съела часть запятых, но в остальном, она показывает хороший результат

| Jamspell | DeepP | YSpeller |
| --- | --- | --- |
|0.09670400619506836 | 0.3538050651550293 | 82.77911901473999 |

Написать функцию Дамерау- Левенштейна

In [47]:
def damerau_levenshtein_distance(s1, s2):
    d = {}
    lenstr1 = len(s1)
    lenstr2 = len(s2)
    for i in range(-1, lenstr1 + 1):
        d[(i, -1)] = i + 1
    for j in range(-1, lenstr2 + 1):
        d[(-1, j)] = j + 1

    for i in range(lenstr1):
        for j in range(lenstr2):
            if s1[i] == s2[j]:
                cost = 0
            else:
                cost = 1
            d[(i, j)] = min(
                d[(i - 1, j)] + 1,  # удаление
                d[(i, j - 1)] + 1,  # вставка
                d[(i - 1, j - 1)] + cost,  # замена
            )
            if i and j and s1[i] == s2[j - 1] and s1[i - 1] == s2[j]:
                d[(i, j)] = min(d[(i, j)], d[i - 2, j - 2] + cost)  # транспозиция

    return d[lenstr1 - 1, lenstr2 - 1]

In [48]:
def correct_text(text, dictionary):
    words = text.split()
    corrected_words = []
    for word in words:
        min_distance = float('inf')
        closest_word = word
        for candidate in dictionary:
            distance = damerau_levenshtein_distance(word, candidate)
            if distance < min_distance:
                min_distance = distance
                closest_word = candidate
        corrected_words.append(closest_word)
    return " ".join(corrected_words)

In [52]:
import json
dictionary = {

  "Универсального": 1,
  "рецепта": 1,
  "того": 1,
  "как": 1,
  "выбрать": 1,
  "правильный": 1,
  "единственно": 1,
  "верный": 1,
  "только": 1,
  "тебе": 1,
  "предназначенный": 1,
  "путь": 1,
  "в": 1,
  "жизни": 1,
  "просто": 1,
  "нет": 1,
  "и": 1,
  "быть": 1,
  "не": 1,
  "может": 1,
  "И": 1,
  "окончательный": 1,
  "выбор": 1,
  "всегда": 1,
  "остается": 1,
  "за": 1,
  "человеком": 1,
  "Этот": 1,
  "мы": 1,
  "делаем": 1,
  "уже": 1,
  "детстве": 1,
  "когда": 1,
  "выбираем": 1,
  "друзей": 1,
  "учимся": 1,
  "строить": 1,
  "отношения": 1,
  "с": 1,
  "ровесниками": 1,
  "играть": 1,
  "Но": 1,
  "большинство": 1,
  "самых": 1,
  "важнейших": 1,
  "решений": 1,
  "определяющих": 1,
  "жизненный": 1,
  "все-таки": 1,
  "принимаем": 1,
  "юности": 1,
  "Как": 1,
  "считают": 1,
  "ученые": 1,
  "вторая": 1,
  "половина": 1,
  "второго": 1,
  "десятилетия": 1,
  "жизни": 1,
  "самый": 1,
  "ответственный": 1,
  "период": 1,
  "Именно": 1,
  "это": 1,
  "время": 1,
  "человек": 1,
  "самое": 1,
  "главное": 1,
  "на": 1,
  "всю": 1,
  "ближайшего": 1,
  "друга": 1,
  "круг": 1,
  "основных": 1,
  "интересов": 1,
  "профессию": 1,
  "Понятно": 1,
  "что": 1,
  "такой": 1,
  "дело": 1,
  "ответственное": 1,
  "От": 1,
  "него": 1,
  "невозможно": 1,
  "отмахнуться": 1,
  "его": 1,
  "нельзя": 1,
  "отложить": 1,
  "потом": 1,
  "Не": 1,
  "стоит": 1,
  "надеяться": 1,
  "ошибку": 1,
  "после": 1,
  "можно": 1,
  "будет": 1,
  "исправить": 1,
  "успеется": 1,
  "вся": 1,
  "впереди": 1,
  "Что-то": 1,
  "конечно": 1,
  "удастся": 1,
  "подправить": 1,
  "изменить": 1,
  "но": 1,
  "далеко": 1,
  "все": 1,
  "неверные": 1,
  "без": 1,
  "последствий": 1,
  "останутся": 1,
  "Ведь": 1,
  "успех": 1,
  "приходит": 1,
  "к": 1,
  "тем": 1,
  "кто": 1,
  "знает": 1,
  "чего": 1,
  "он": 1,
  "хочет": 1,
  "решительно": 1,
  "делает": 1,
  "верит": 1,
  "себя": 1,
  "упорно": 1,
  "достигает": 1,
  "намеченных": 1,
  "целей": 1
}


with open("dictionary.json", "w", encoding="utf-8") as f:
    json.dump(dictionary, f, ensure_ascii=False, indent=2)

In [55]:

try:
    with open("dictionary.json", "r", encoding="utf-8") as f:
        dictionary = json.load(f)
except FileNotFoundError:
    print("dictionary.json not found.  Create a dictionary file.")
    dictionary = {}
timestart = time.time()
corrected_text = correct_text(wrong_text, dictionary)
timeend = time.time()
print(f"time is {timeend - timestart}")
print(f"Исправленный текст: {corrected_text}")

time is 0.6825466156005859
Исправленный текст: Универсального рецепта того как выбрать правильный единственно верный только тебе предназначенный путь в жизни просто нет и быть не может И окончательный выбор всегда остается за человеком Этот выбор мы делаем уже в детстве когда выбираем друзей учимся строить отношения с ровесниками играть Но большинство самых важнейших решений определяющих жизненный путь мы всю как принимаем в юности Как считают ученые вторая половина второго десятилетия жизни в самый ответственный период Именно в это время человек выбираем самое главное и на всю жизни ближайшего друга круг основных интересов профессию Понятно что такой выбор дело ответственное От него невозможно отмахнуться его нельзя отложить на потом Не стоит надеяться что ошибку после можно будет исправить успеется вся жизни впереди Что-то конечно удастся подправить изменить но далеко не всю И не верный решений без последствий не останутся Ведь успех приходит к тем кто знает чего он хочет решительно 

Не совсем понял. что имеллось в виду, но решено к каждой букве из алфавита сопоставить в буквы вокруг ней

In [64]:
import random
mistakes = {
    'й': ['ц', 'ф'],
    'ц': ['й', 'у', 'ф', 'ы'],
    'у': ['ц', 'к', 'ы', 'в'],
    'к': ['у', 'е', 'в', 'а'],
    'е': ['к', 'н', 'а', 'п'],
    'н': ['е', 'г', 'п', 'р'],
    'г': ['н', 'ш', 'р', 'о'],
    'ш': ['г', 'щ', 'о', 'л'],
    'щ': ['ш', 'з', 'л', 'д'],
    'з': ['щ', 'х', 'д', 'ж'],
    'х': ['з', 'ъ', 'ж', 'э'],
    'ъ': ['х', 'э'],
    'ф': ['й', 'ц', 'ы', 'я'],
    'ы': ['ф', 'ц', 'у', 'я', 'ч'],
    'в': ['у', 'к', 'а', 'ч', 'с'],
    'а': ['к', 'е', 'в', 'с', 'м'],
    'п': ['е', 'н', 'р', 'м', 'и'],
    'р': ['н', 'г', 'п', 'и', 'т'],
    'о': ['г', 'ш', 'л', 'т', 'ь'],
    'л': ['ш', 'щ', 'д', 'ь', 'б'],
    'д': ['щ', 'з', 'ж', 'б', 'ю'],
    'ж': ['з', 'х', 'э', 'ю'],
    'э': ['х', 'ъ', 'ж'],
    'я': ['ф', 'ы', 'ч'],
    'ч': ['ы', 'в', 'я', 'с'],
    'с': ['в', 'а', 'ч', 'м'],
    'м': ['а', 'п', 'с', 'и'],
    'и': ['п', 'р', 'м', 'т'],
    'т': ['р', 'о', 'и', 'ь'],
    'ь': ['о', 'л', 'т', 'б'],
    'б': ['л', 'д', 'ь', 'ю'],
    'ю': ['д', 'ж', 'б']
}

In [57]:
# ну нужно написать функцию генерации, давайте сделаем это
# идея будет в том, что мы с 0.1 вероятностью берём символ и заменяем его на

In [67]:
def generate_typos(word, typo_prob=0.1, mistakes=mistakes):
    word_with_typos = ""
    for char in word:
        if random.random() < typo_prob and char.lower() in mistakes:
            possible_typos = mistakes[char.lower()]
            typo = random.choice(possible_typos)
            word_with_typos += typo if char.islower() else typo.upper() # Сохраняем регистр
        else:
            word_with_typos += char
    return word_with_typos


# Пример использования:
word = "Привет"
word_with_typos = generate_typos(word)
print(f"Исходное слово: {word}")
print(f"Слово с опечатками: {word_with_typos}")

Исходное слово: Привет
Слово с опечатками: Привет


Задача 4

In [71]:
import numpy as np
from collections import Counter

def CountVectorizer(word1, word2):

    if not word1 or not word2:
        return 0
    word1 = word1.lower()
    word2 = word2.lower()

    counter1 = Counter(word1)
    counter2 = Counter(word2)

    all_chars = set(counter1.keys() | counter2.keys())

    vector1 = np.array([counter1.get(char, 0) for char in all_chars])
    vector2 = np.array([counter2.get(char, 0) for char in all_chars])

    # Вычисляем косинусное сходство
    cosine_similarity = np.dot(vector1, vector2) / (np.linalg.norm(vector1) * np.linalg.norm(vector2))

    return cosine_similarity



print(compare_letter_frequency_vectors("привет", "привет"))   # Вывод: 1.0
print(compare_letter_frequency_vectors("привет", "ПРИВЕТ"))   # Вывод: 1.0 (регистр не учитывается)
print(compare_letter_frequency_vectors("привет", "превед"))   # Вывод: ~0.82
print(compare_letter_frequency_vectors("привет", "пока"))     # Вывод: ~0.27
print(compare_letter_frequency_vectors("привет", ""))        # Вывод: 0
print(compare_letter_frequency_vectors("", "пока"))

1.0000000000000002
1.0000000000000002
0.7216878364870323
0.20412414523193154
0
0


In [72]:
  def correct_text(text, dictionary, typo_prob = 0.2, keyboard_layout = mistakes):
    words = text.split()
    corrected_words = []
    for word in words:
        if word in dictionary:  # Слово уже есть в словаре - не нужно исправлять
            corrected_words.append(word)
            continue


        best_candidate = None
        max_similarity = 0


        candidates = [candidate for candidate in dictionary if abs(len(candidate) - len(word)) <= 2 ]

        for candidate in candidates:
            similarity = CountVectorizer(word, candidate)
            if similarity > max_similarity and damerau_levenshtein_distance(word.lower(), candidate.lower()) <= len(word) * typo_prob :
                max_similarity = similarity
                best_candidate = candidate

        if best_candidate:
            corrected_words.append(best_candidate)
        else:

            min_distance = float('inf')
            closest_word = word

            for candidate in dictionary:
                distance = damerau_levenshtein_distance(word.lower(), candidate.lower())
                if distance < min_distance:
                    min_distance = distance
                    closest_word = candidate

            corrected_words.append(closest_word)

    return " ".join(corrected_words)


In [73]:
timestart = time.time()
corrected_text = correct_text(wrong_text, dictionary)
timeend = time.time()
print(f"time is {timeend - timestart}")
print(f"Исправленный текст: {corrected_text}")

time is 0.19004392623901367
Исправленный текст: Универсального рецепта того как выбрать правильный единственно верный только тебе предназначенный путь в жизни просто нет и быть не может И окончательный выбор всегда остается за человеком Этот выбор мы делаем уже в детстве когда выбираем друзей учимся строить отношения с ровесниками играть Но большинство самых важнейших решений определяющих жизненный путь мы всю как принимаем в юности Как считают ученые вторая половина второго десятилетия жизни в самый ответственный период Именно в это время человек выбираем самое главное и на всю жизни ближайшего друга круг основных интересов профессию Понятно что такой выбор дело ответственное От него невозможно отмахнуться его нельзя отложить на потом Не стоит надеяться что ошибку после можно будет исправить успеется вся жизни впереди Что-то конечно удастся подправить изменить но далеко не всю И не верный решений без последствий не останутся Ведь успех приходит к тем кто знает чего он хочет решительно

|время без проверки косинусной близости | время с проверкой косинусной близовтси |
|--- | ---- |
| 0.6825466156005859 | 0.19004392623901367 |

Итог, существуют разные методы для удаления ошибок в тексте, такие как yandexSpeller и другие

Имеются базовые средства для исключения ошибок, такие как метод Дамерау-  Левенштейна, которые можно ускорить, если использовать косинусную близость между данными векторами